quadriga/src/main/java/edu/asu/diging/quadriga/core/service/impl/MappedTripleGroupServiceImpl.java

```diff
@@ -0,0 +1,151 @@
+import edu.asu.diging.quadriga.core.model.MappedTripleGroup;
+import edu.asu.diging.quadriga.core.model.MappedTripleType;
+import edu.asu.diging.quadriga.core.service.CollectionManager;
+import edu.asu.diging.quadriga.core.service.MappedTripleGroupService;
+
+@Service
+public class MappedTripleGroupServiceImpl implements MappedTripleGroupService {
+    
+    private Logger logger = LoggerFactory.getLogger(getClass());
+
+    @Autowired
+    private MappedTripleGroupRepository mappedTripleGroupRepository;
+
+    @Autowired
+    private CollectionManager collectionManager;
+
+    /**
+     * Converts the given collectionId into an ObjectId and persists a
+     * MappedTripleGroup in the database with this collectionId and MappedTripleType.
+     * 
+     * @param collectionId set to the new MappedTripleGroup
+     * @param mappedTripleType is the group type which the current triple belongs to
+     * @return the persisted MappedTripleGroup object
+     * @throws InvalidObjectIdException    if collectionId couldn't be converted to
+     *                                     ObjectId
+     * @throws CollectionNotFoundException if collection with given collectionId
+     *                                     does't exist
+     */
+    @Override
+    public MappedTripleGroup add(String collectionId, MappedTripleType mappedTripleType)
+            throws InvalidObjectIdException, CollectionNotFoundException {
+        
+        MappedTripleGroup mappedTripleGroup = new MappedTripleGroup();
+        
+        mappedTripleGroup.setCollectionId(collectionManager.getCollection(collectionId).getId());
+        mappedTripleGroup.setMappedTripleType(mappedTripleType);
+        
+        return mappedTripleGroupRepository.save(mappedTripleGroup);
+    }
+
+    /**
+     * This method first checks whether a collection with the given collectionId
+     * exists.
+     * 
+     * If yes, it finds a MappedTripleGroup entry in the database with this
+     * collectionId and the MappedTripleType.
+     * 
+     * @param collectionId used to look for the MappedTripleGroup entry
+     * @param mappedTripleType is the group type to be looked for in the database
+     * @return the MappedTripleGroup entry that was found, else null
+     * @throws InvalidObjectIdException    if collectionId couldn't be converted to
+     *                                     ObjectId
+     * @throws CollectionNotFoundException if collection with given collectionId
+     *                                     doesn't exist
+     */
+    @Override
+    public MappedTripleGroup findByCollectionIdAndMappingType(String collectionId, MappedTripleType mappedTripleType)
+            throws InvalidObjectIdException, CollectionNotFoundException {
+        return mappedTripleGroupRepository
+                .findByCollectionIdAndMappedTripleType(collectionManager
+                        .getCollection(collectionId)
+                        .getId(), mappedTripleType)
```

Comment:
Why are you retrieving the collection here? Why can't you just create an objectId out of the colletionId directly?

Reply:
After a network is submitted for a collection, two scenarios could occur -

The collection exists, and it's the first time a network is being submitted to it
The collection doesn't exist, i.e. the network submission request is incorrect
In the first case, we add new mappedTripleGroup to the existing collection
In the second case, we return null, which will send a NOT_FOUND response to the request

Now here, if we don't retrieve the collection and simply pass new ObjectId(collectionId), this method will return null for both the above cases
To distinguish between the two, we need to retrieve the collection here

I can change this method to -
Try to get Collection, and if it doesn't exist, throw CollectionNotFoundException here
If it exists, null will be returned, which means we need to add new MappedTripleGroup

Is that fine?

Reply:
Something like this -
```java
@Override
public MappedTripleGroup findByCollectionIdAndMappingType(String collectionId, MappedTripleType mappedTripleType)
    throws InvalidObjectIdException, CollectionNotFoundException {
    Collection collection = collectionManager.findCollection(collectionId);
        
    if(collection == null) {
        throw new CollectionNotFoundException("Couldn't find collection for id: " + collectionId);
    }
        
    return mappedTripleGroupRepository
            .findByCollectionIdAndMappedTripleType(collection.getId(), mappedTripleType)
            .orElse(null);
}
```

Reply:
I"m confused. This method is called `findByCollectionIdAndMappingType` which indicates it simply returns a collection with the given id and mapping type. I expect this method to not return anything if the id and mapping type doesn't exist. What the caller does with that info is a different story. They can decide to create a new collection or raise a not found. I would make it the job of the caller to ensure the collection exists, but if you want to keep the check here and then, the above solution is better.

Reply:
Yes, this method does return `null` if the collectionId + mapping type doesn't exist
The part where new MappedTripleEntry is created happens in the calling method, not here
I'll keep this logic and re-open the PR, let me know if that looks correct